pyomo/core/tests/unit/test_taylor_series.py

```diff
@@ -1,17 +1,47 @@
 import pyutilib.th as unittest
 import pyomo.environ as pe
-from pyomo.core.expr.taylor_series import get_taylor_series
+from pyomo.core.expr.taylor_series import taylor_series_expansion
 from pyomo.core.expr.current import polynomial_degree
+from pyomo.core.expr.calculus.derivatives import differentiate
 
 
-class TestDerivs(unittest.TestCase):
-    def test_taylor_series(self):
+class TestTaylorSeries(unittest.TestCase):
+    def test_first_order_taylor_series(self):
         m = pe.ConcreteModel()
         m.x = pe.Var()
         m.x.value = 1
         exprs_to_test = [m.x**2, pe.exp(m.x), (m.x + 2)**2]
         for e in exprs_to_test:
-            tsa = get_taylor_series(e)
+            tsa = taylor_series_expansion(e)
             self.assertAlmostEqual(pe.differentiate(e, wrt=m.x), pe.differentiate(tsa, wrt=m.x))
             self.assertAlmostEqual(pe.value(e), pe.value(tsa))
             self.assertEqual(polynomial_degree(tsa), 1)
+
+    def test_higher_order_taylor_series(self):
+        m = pe.ConcreteModel()
+        m.x = pe.Var(initialize=0.5)
+        m.y = pe.Var(initialize=1.5)
+
+        e = m.x * m.y
+        tse = taylor_series_expansion(e, diff_mode=differentiate.Modes.reverse_symbolic, order=2)
+        for _x in [-2, -1.5, -1, -0.5, 0, 0.5, 1, 1.5, 2]:
+            for _y in [-2, -1.5, -1, -0.5, 0, 0.5, 1, 1.5, 2]:
+                m.x.value = _x
+                m.y.value = _y
+                self.assertAlmostEqual(pe.value(e), pe.value(tse))
+
+        e = m.x**3 + m.y**3
+        tse = taylor_series_expansion(e, diff_mode=differentiate.Modes.reverse_symbolic, order=3)
+        for _x in [-2, -1.5, -1, -0.5, 0, 0.5, 1, 1.5, 2]:
+            for _y in [-2, -1.5, -1, -0.5, 0, 0.5, 1, 1.5, 2]:
+                m.x.value = _x
+                m.y.value = _y
+                self.assertAlmostEqual(pe.value(e), pe.value(tse))
+
+        e = (m.x*m.y)**2
+        tse = taylor_series_expansion(e, diff_mode=differentiate.Modes.reverse_symbolic, order=4)
```

Comment:
Does this test make sense?  The 3rd and 4th order terms should be 0...

I think I would be generally more comfortable if these tests ensured that the correct number of terms were being returned and that the individual values in the expansion were correct.

Reply:
If the expression were ```(m.x + m.y)**2```, then the 3rd and 4th order terms would be zero. As written, this is a 4th order polynomial. Thus, the 4th order taylor series expansion should be exact, which is essentially what this test checks. In fact, that is what all of these tests are doing. This is why I focused on testing polynomials. With polynomials, we can test that a taylor series of the same order as the polynomial returns the same polynomial. It is not terribly easy to test that the expressions generated are exactly what I expect without writing a custom walker. However, it is very unlikely that the value of the original expression and the value of the taylor series expansion will be identical at all of the points being tested unless the expressions are identical. In my opinion, that makes these very good tests.

Keep in mind that, in the current implemenation, the higher order taylor series expansions of multivariate expressions are very ugly. This is because the implementation is rather naive, and no simplification is done. For example, the 4th order taylor series expansion for 

```(m.x*m.y)**2``` 

is 

```2.25*(x - 0.5) + 0.75*(y - 1.5) + 0.5625 + 0.5*(x - 0.5)*(4.5*(x - 0.5) + 3.0*(y - 1.5)) + 0.5*(y - 1.5)*(3.0*(x - 0.5) + 0.5*(y - 1.5)) + 0.16666666666666666*(x - 0.5)*(x - 0.5)*(6.0*(y - 1.5)) + 0.16666666666666666*(x - 0.5)*(y - 1.5)*(6.0*(x - 0.5) + 2.0*(y - 1.5)) + 0.16666666666666666*(y - 1.5)*(x - 0.5)*(6.0*(x - 0.5) + 2.0*(y - 1.5)) + 0.16666666666666666*(y - 1.5)*(y - 1.5)*(2.0*(x - 0.5)) + 0.041666666666666664*(x - 0.5)*(x - 0.5)*(y - 1.5)*(4*(y - 1.5)) + 0.041666666666666664*(x - 0.5)*(y - 1.5)*(x - 0.5)*(4*(y - 1.5)) + 0.041666666666666664*(x - 0.5)*(y - 1.5)*(y - 1.5)*(4*(x - 0.5)) + 0.041666666666666664*(y - 1.5)*(x - 0.5)*(x - 0.5)*(4*(y - 1.5)) + 0.041666666666666664*(y - 1.5)*(x - 0.5)*(y - 1.5)*(4*(x - 0.5)) + 0.041666666666666664*(y - 1.5)*(y - 1.5)*(x - 0.5)*(4*(x - 0.5))```

 If you simplify it, it does turn out to be ```x**2*y**2```.